pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -48,33 +55,112 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/yaml"
 )
 
 const (
 	// This controller manages kubeone external cluster.
 	ControllerName = "kubeone_controller"
 	ImportAction   = "import"
+	// Upgrade control plane only.
+	UpgradeControlPlaneAction = "upgradeCP"
+	// KubeOne Pod Name.
+	KubeOneUpgradePod    = "kubeone-upgrade"
+	KubeOneKubeconfigPod = "kubeone-kubeconfig"
 )
 
 type reconciler struct {
 	ctrlruntimeclient.Client
 	log *zap.SugaredLogger
+	kubernetesprovider.ImpersonationClient
+	secretKeySelector provider.SecretKeySelectorValueFunc
 }
 
 func Add(
 	ctx context.Context,
 	mgr manager.Manager,
 	log *zap.SugaredLogger) error {
+	defaultImpersonationClient := kubernetesprovider.NewImpersonationClient(ctrlruntime.GetConfigOrDie(), mgr.GetRESTMapper())
```

Comment:
I don't think a controller should use GetConfigOrDie(), that function is for main()'s. Please rely only on parameters to your Add() function instead of performing such initialization work in here.

Reply:
Good Point